sql: postgres jdbc doesn't work with BACKUP

The following program panics CockroachDB with the `org.postgresql:postgresql:9.4.1212` jar, but the statement works fine from both `cockroach sql` and `psql`.

```
import java.sql.*;

public class Test {
    public static void main(String[] args) throws Exception {
        Class.forName("org.postgresql.Driver");
        Connection conn = null;
        try {
            conn = DriverManager.getConnection("jdbc:postgresql://127.0.0.1:26257/system?sslmode=disable", "root", "");
            conn.createStatement().executeQuery("BACKUP DATABASE system TO '/tmp/foo'");
        } finally {
            if (conn != null) { conn.close(); }
        }
    }
}
```

logs from CockroachDB
```
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 598 [running]:
panic(0x52edc80, 0xc420012110)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4203a63f0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x52edc80, 0xc420012110)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420413600, 0x7c9a980, 0xc420a754d0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:340 +0x74
panic(0x52edc80, 0xc420012110)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).sendResponse(0xc420413600, 0x7c9a980, 0xc420a754d0, 0xc4208dd9f0, 0x1, 0x1, 0x62cae40, 0x0, 0x0, 0x531c600, ...)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:953 +0x6f9
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc420413600, 0x7c9a980, 0xc420a754d0, 0xc42085702c, 0x24, 0xc420e2b470, 0x62cae40, 0x0, 0x0, 0xc42016fa00, ...)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:822 +0x202
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleExecute(0xc420413600, 0x7c9a980, 0xc420a754d0, 0xc420413628, 0x9, 0x0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:799 +0x20a
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420413600, 0x7c9a980, 0xc420a754d0, 0xc420b259b0, 0x5400, 0xc420328138, 0x7c9a980, 0xc420a754a0, 0x0, 0x0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:442 +0xaea
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc420328000, 0x7c9a980, 0xc420a754a0, 0x5da64c0, 0xc420a653f0, 0x0, 0x0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:308 +0xa22
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x5da64c0, 0xc420a653f0)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/server/server.go:527 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4203a63f0, 0xc420136130, 0x5da64c0, 0xc420a653f0, 0xc420ab0000)
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/Users/dan/Dropbox/Projects/crl1/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

--- a/pkg/sql/plan.go
+++ b/pkg/sql/plan.go
@@ -17,6 +17,8 @@
 package sql
 
 import (
+	"golang.org/x/net/context"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
@@ -220,6 +222,24 @@ func (p *planner) startPlan(plan planNode) error {
 	return nil
 }
 
+func maybePlanHook(
+	ctx context.Context, stmt parser.Statement, cfg *ExecutorConfig,
+) (planNode, error) {
+	// TODO(dan): This iteration makes the plan dispatch no longer constant
+	// time. We could fix that with a map of `reflect.Type` but including
+	// reflection in such a primary codepath is unfortunate. Instead, the
+	// upcoming IR work will provide unique numeric type tags, which will
+	// elegantly solve this.
+	for _, planHook := range planHooks {
+		if fn, header, err := planHook(ctx, stmt, cfg); err != nil {
+			return nil, err
+		} else if fn != nil {
+			return &hookFnNode{f: fn, header: header}, nil
+		}
+	}
+	return nil, nil
+}
+
 // newPlan constructs a planNode from a statement. This is used
 // recursively by the various node constructors.
 func (p *planner) newPlan(
@@ -236,17 +256,8 @@ func (p *planner) newPlan(
 		p.txn.SetSystemConfigTrigger()
 	}
 
-	// TODO(dan): This iteration makes the plan dispatch no longer constant
-	// time. We could fix that with a map of `reflect.Type` but including
-	// reflection in such a primary codepath is unfortunate. Instead, the
-	// upcoming IR work will provide unique numeric type tags, which will
-	// elegantly solve this.
-	for _, planHook := range planHooks {
-		if fn, header, err := planHook(p.ctx(), stmt, p.execCfg); err != nil {
-			return nil, err
-		} else if fn != nil {
-			return &hookFnNode{f: fn, header: header}, nil
-		}
+	if plan, err := maybePlanHook(p.ctx(), stmt, p.execCfg); plan != nil || err != nil {
+		return plan, err
 	}
 
 	switch n := stmt.(type) {
@@ -346,6 +357,10 @@ func (p *planner) newPlan(
 }
 
 func (p *planner) prepare(stmt parser.Statement) (planNode, error) {
+	if plan, err := maybePlanHook(p.ctx(), stmt, p.execCfg); plan != nil || err != nil {
+		return plan, err
+	}
+
 	switch n := stmt.(type) {
 	case *parser.Delete:
 		return p.Delete(n, nil, false)

--- a/pkg/sql/planhook_test.go
+++ b/pkg/sql/planhook_test.go
@@ -1,0 +1,45 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Daniel Harrison (daniel.harrison@gmail.com)
+
+package sql_test
+
+import (
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+func TestPlanHookPrepare(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{})
+	defer tc.Stopper().Stop()
+	sqlDB := tc.Conns[0]
+
+	stmt, err := sqlDB.Prepare(`SHOW planhook`)
+	if err != nil {
+		t.Fatalf("%+v", err)
+	}
+	row := stmt.QueryRow()
+	var s string
+	if err := row.Scan(&s); err != nil {
+		t.Fatalf("%+v", err)
+	}
+	if s != `planhook` {
+		t.Errorf("got %s expected planhook", s)
+	}
+}

